src/Compilers/Core/Portable/Text/SubText.cs

```diff
@@ -89,5 +94,87 @@ private TextSpan GetCompositeSpan(int start, int length)
             int compositeEnd = Math.Min(UnderlyingText.Length, compositeStart + length);
             return new TextSpan(compositeStart, compositeEnd - compositeStart);
         }
+
+        /// <summary>
+        /// Delegates to the SubText's UnderlyingText to determine line information.
+        /// </summary>
+        private sealed class SubTextLineInfo : TextLineCollection
+        {
+            private readonly SubText _subText;
+            private readonly int _startLineNumber;
+            private readonly int _lineCount;
+            private readonly bool _startsWithinSplitCRLF;
+            private readonly bool _endsWithinSplitCRLF;
+
+            public SubTextLineInfo(SubText subText)
+            {
+                _subText = subText;
+
+                var startLine = _subText.UnderlyingText.Lines.GetLineFromPosition(_subText.UnderlyingSpan.Start);
+                var endLine = _subText.UnderlyingText.Lines.GetLineFromPosition(_subText.UnderlyingSpan.End);
+
+                _startLineNumber = startLine.LineNumber;
+                _lineCount = (endLine.LineNumber - _startLineNumber) + 1;
+
+                if (_subText.UnderlyingSpan.Length > 0)
+                {
+                    if (_subText.UnderlyingSpan.Start == startLine.End + 1 &&
+                        _subText.UnderlyingSpan.Start == startLine.EndIncludingLineBreak - 1)
+                    {
+                        _startsWithinSplitCRLF = true;
+                    }
+
+                    if (_subText.UnderlyingSpan.End == endLine.End + 1 &&
+                        _subText.UnderlyingSpan.End == endLine.EndIncludingLineBreak - 1)
+                    {
+                        _endsWithinSplitCRLF = true;
+
+                        // If this subtext ends in the middle of a CR/LF, then this object should view that CR as a separate line
+                        // whereas the UnderlyingText would not.
+                        _lineCount += 1;
+                    }
+                }
+            }
+
+            public override TextLine this[int lineNumber]
+            {
+                get
+                {
+                    if (_endsWithinSplitCRLF && lineNumber == _lineCount - 1)
+                    {
+                        // Special case splitting the CRLF at the end as the UnderlyingText doesn't view the position
+                        // after between the \r and \n as on a new line whereas this subtext doesn't contain the \n
+                        // and needs to view that position as on a new line.
+                        return TextLine.FromSpanUnsafe(_subText, new TextSpan(_subText.UnderlyingSpan.End, 0));
+                    }
+
+                    var underlyingTextLine = _subText.UnderlyingText.Lines[lineNumber + _startLineNumber];
+
+                    var startInUnderlyingText = Math.Max(underlyingTextLine.Start, _subText.UnderlyingSpan.Start);
```

Comment:
not understanding how this coudl ever pick the first.  need comment/example.

Reply:
The table itself makes sense to me, but I don't understand the meaning/purpose of `startInUnderlyingText`. Will revisit on Monday and complete review. Thanks!

Reply:
It represents the position of what we view as the start of the requested line, but in the coordinate space of _subText.UnderlyingText.

Reply:
a comment would be helpful.

Reply:
I think I get this now. In the case that `_subText.UnderlyingSpan.Start` is bigger, it means the subtext starts on the middle of a line in the underlying text. And if the `underlyingTextLine.Start` is bigger, it means we are looking at a subsequent line, i.e. some line after line 0 of the `_subText`. And pretty much the same dynamic is going on with the other variable, except relating to the ends of lines and the last line of the `_subText`.